### ISSUE TITLE
Template variables making dashboards slow to load since 3.0.4

### ISSUE THREAD

**Issue by benthorner** [grafana/grafana#5637]
* **I'm submitting a ... Bug report

- What Grafana version are you using?
      3.0.3, 3.0.4
- What datasource are you using?
      graphite
- What OS are you running grafana on?
      RHEL 6.4
- What did you do?
      Created a dashboard with a hierarchy of template variables to select between graphs. Each variable is set to refresh on dashboard load; the range of values of the variables grows exponentially (up to 100s).
- What was the expected result?
      The dashboard loads immediately in Grafana 3.0.3, but takes up to 30 seconds in 3.0.4 and above; this behaviour is also observed to a lesser extent in other dashboards (where the range of values and number of variables is smaller).
- What happened instead?
      Looking at the network requests, I can see that 3.0.3, each variable corresponds to a single backend query when the page loads. In 3.0.4, each variable corresponds to many repeated queries, using the saved values of the variables, followed by multiple queries with values overridden in the URL params. The number of repeated queries seems to lock-up the grafana backend, causing the page to take a long time to load.

**Comment by torkelo** [comment]
Yes, in Grafana 3.0.4, variables that are marked on refresh on load will block the loading of the dashboard. The variable values need to be fetched before the dashboard can load. 

If you do not want this behavior you can change the template variables refresh property, or try to improve the performance of your variable value queries.

**Comment by benthorner** [comment]
Hi @torkelo. Yes, variables marked refresh on load should block the the loading of the dashboard. The problem is the variable values are being fetched *more than once* for the same variable.

For example, let's say we have the following variables:

  a: root.*
  b: root.$a.*
  c: root.$a.$b.*

Since 3.0.4, we observe that $a is fetched multiple times, $b at least twice as much as $a, and $c at least twice as much as $b i.e. an exponentially increasing number of queries.

Looking at the code, I can see at least one culprit: the *validateVariableSelectionState* function does not check for *initPhase*. This alone will cause an exponential chain of variable updates. 

Our datasource is pretty fast, but this exponential number of queries is causing a bottleneck. I think this issue should be re-opened - making repeated backend queries is unnecessary.

**Comment by torkelo** [comment]
you right! will fix!

**Comment by benthorner** [comment]
Thanks @torkelo. I forgot to say that this will also happen when template variables are updated normally i.e. any call to updateOptionsInChildVariables. For example, say we have

   a: root.*
   b: root.$a.*
   c: root.$a.$b.*

Then a change to $a will cause 1 query for $b, but 2 queries for $c (the second query is because $b is updated). It's kind of hard to think about the control flow, but I think this adds to the problem.

Perhaps a fix would be to somehow move/centralise the dependency check logic in init/processVariables to the updateOptionsInChildVariables method. Hope that helps!
